libraries/ms-common/src/x86/adc.c

```diff
@@ -176,32 +178,200 @@ StatusCode adc_read_converted(AdcChannel adc_channel, uint16_t *reading) {
+    case GPIO_PORT_A:
+      if (address.pin > 7) {
+        return status_code(STATUS_CODE_INVALID_ARGS);
+      }
+      break;
+    case GPIO_PORT_B:
+      if (address.pin > 1) {
+        return status_code(STATUS_CODE_INVALID_ARGS);
+      }
+      *adc_channel += 8;
+      break;
+    case GPIO_PORT_C:
+      if (address.pin > 5) {
+        return status_code(STATUS_CODE_INVALID_ARGS);
+      }
+      *adc_channel += 10;
+      break;
+  }
+
+  if (*adc_channel > ADC_CHANNEL_15) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode adc_register_callback(AdcChannel adc_channel, AdcCallback callback,
+                                 void *context) {  // no change?
+  if (adc_channel >= NUM_ADC_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  if (s_active_channels[adc_channel] != true) {
+    return status_code(STATUS_CODE_EMPTY);
+  }
+
+  s_adc_interrupts[adc_channel].callback = callback;
   s_adc_interrupts[adc_channel].context = context;
+
   return STATUS_CODE_OK;
 }
 
-StatusCode adc_read_raw_pin(GpioAddress address, uint16_t *reading) {
-  AdcChannel channel;
-  status_ok_or_return(adc_get_channel(address, &channel));
-  return adc_read_raw(channel, reading);
+StatusCode adc_read_raw(AdcChannel adc_channel, uint16_t *reading) {
+  if (adc_channel >= NUM_ADC_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  if (s_active_channels[adc_channel] != true) {
+    return status_code(STATUS_CODE_EMPTY);
+  }
+
+  s_adc_interrupts[adc_channel].reading = ADC_RETURNED_VOLTAGE_RAW;
+  s_store.reading[adc_channel] = ADC_RETURNED_VOLTAGE_RAW;
+  *reading = s_adc_interrupts[adc_channel].reading;
+
+  if (s_adc_interrupts[adc_channel].callback != NULL) {
+    s_adc_interrupts[adc_channel].callback(adc_channel, s_adc_interrupts[adc_channel].context);
+  }
+  prv_export();
+
+  return STATUS_CODE_OK;
 }
 
-StatusCode adc_read_converted_pin(GpioAddress address, uint16_t *reading) {
-  AdcChannel adc_channel;
-  status_ok_or_return(adc_get_channel(address, &adc_channel));
-  return adc_read_converted(adc_channel, reading);
+StatusCode adc_read_converted(AdcChannel adc_channel, uint16_t *reading) {
+  if (adc_channel >= NUM_ADC_CHANNELS) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+  if (s_active_channels[adc_channel] != true) {
+    return status_code(STATUS_CODE_EMPTY);
+  }
+
+  uint16_t adc_reading = 0;
+  adc_read_raw(adc_channel, &adc_reading);
+  switch (adc_channel) {
+    case ADC_CHANNEL_TEMP:
+      *reading = prv_get_temp(adc_reading);
+      return STATUS_CODE_OK;
+
+    case ADC_CHANNEL_REF:
+      *reading = prv_get_vdda(adc_reading);
+      return STATUS_CODE_OK;
+
+    case ADC_CHANNEL_BAT:
+      adc_reading *= 2;
+      break;
+
+    default:
+      break;
+  }
+
+  uint16_t vdda;
+  adc_read_converted(ADC_CHANNEL_REF, &vdda);
+  *reading = (adc_reading * vdda) / 4095;
+
+  s_store.reading[adc_channel] = *reading;
```

Comment:
We shouldn't be setting the reading here. This function reads a value from the hardware, but the hardware is simulated by python, so setting the value here would be overwriting the value set by python. Instead we should just return the value set by python as in `return s_store.reading[adc_channel]`